pkg/app/pipedv1/plugin/kubernetes/provider/resource.go

```diff
@@ -29,42 +31,101 @@ const (
 )
 
 type ResourceKey struct {
-	apiVersion string
-	kind       string
-	namespace  string
-	name       string
+	groupKind schema.GroupKind
+	namespace string
+	name      string
 }
 
-func (k ResourceKey) APIVersion() string {
-	return k.apiVersion
+func (k ResourceKey) Kind() string {
+	return k.groupKind.Kind
 }
 
-func (k ResourceKey) Kind() string {
-	return k.kind
+func (k ResourceKey) Name() string {
+	return k.name
 }
 
 func (k ResourceKey) Namespace() string {
 	return k.namespace
 }
 
-func (k ResourceKey) Name() string {
-	return k.name
+// normalize converts the group and kind to lower case.
+// It also converts the default namespace to an empty string.
+func (k ResourceKey) normalize() ResourceKey {
+	k.groupKind = normalizeGroupKind(k.groupKind)
+	return k.normalizeNamespace()
+}
+
+// normalizeNamespace converts the default namespace to an empty string.
+func (k ResourceKey) normalizeNamespace() ResourceKey {
+	if k.namespace == DefaultNamespace {
+		return k.withoutNamespace()
+	}
+	return k
+}
+
+func (k ResourceKey) withoutNamespace() ResourceKey {
+	k.namespace = ""
+	return k
 }
 
 func (k ResourceKey) String() string {
-	return fmt.Sprintf("%s:%s:%s:%s", k.apiVersion, k.kind, k.namespace, k.name)
+	return fmt.Sprintf("%s:%s:%s:%s", k.groupKind.Group, k.groupKind.Kind, k.namespace, k.name)
 }
 
 func (k ResourceKey) ReadableString() string {
-	return fmt.Sprintf("name=%q, kind=%q, namespace=%q, apiVersion=%q", k.name, k.kind, k.namespace, k.apiVersion)
+	return fmt.Sprintf("name=%q, kind=%q, namespace=%q, apiGroup=%q", k.name, k.groupKind.Kind, k.namespace, k.groupKind.Group)
 }
 
 func makeResourceKey(obj *unstructured.Unstructured) ResourceKey {
 	k := ResourceKey{
-		apiVersion: obj.GetAPIVersion(),
-		kind:       obj.GetKind(),
-		namespace:  obj.GetNamespace(),
-		name:       obj.GetName(),
+		groupKind: obj.GroupVersionKind().GroupKind(),
+		namespace: obj.GetNamespace(),
+		name:      obj.GetName(),
 	}
 	return k
 }
+
+func normalizeGroupKind(gk schema.GroupKind) schema.GroupKind {
+	gk.Group = strings.ToLower(gk.Group)
+	gk.Kind = strings.ToLower(gk.Kind)
+	return gk
+}
```

Comment:
[imo] How about unifying to `normalize` to avoid using only this method in other code?

Reply:
applied on this commit.
https://github.com/pipe-cd/pipecd/pull/5455/commits/bcfd121f95c074895e5a1986232b54501574e3fc